ykrt/src/thread_intercept.rs

```diff
@@ -1,11 +1,83 @@
 use std::os::raw::{c_int, c_void};
 
+use std::{ffi::CString, ptr::null_mut};
+
+use libc::{dlsym, free, malloc, pthread_attr_t, pthread_create, pthread_t};
+
+const SHADOW_STACK_SIZE: usize = 1000000;
+
+#[derive(Debug)]
+struct ThreadRoutine {
+    pub args: *mut c_void,
+    pub start_routine: extern "C" fn(*mut c_void) -> *mut c_void,
+}
+
+extern "C" fn wrap_thread_routine(arg: *mut c_void) -> *mut c_void {
+    let str = CString::new("shadowstack_0").unwrap();
+    let shadowstack_symbol_addr = unsafe {
+        //  Obtain address of a shadowstack_0 symbol
+        dlsym(null_mut(), str.as_ptr() as *const i8)
+    };
+    if shadowstack_symbol_addr.is_null() {
+        panic!("Unable to find shadowstack address")
+    }
+    let stack_addr = unsafe {
+        // Allocate stack
+        malloc(SHADOW_STACK_SIZE)
+    };
+    if stack_addr.is_null() {
+        panic!("Unable allocate stack")
+    }
+    unsafe {
+        // Set shadowstack symbol with new allocated stack
+        *(shadowstack_symbol_addr as *mut *mut c_void) = stack_addr;
+    }
+    let thread_routine = unsafe {
+        // Obtain ThreadRoutine struct
+        Box::from_raw(arg as *mut ThreadRoutine)
+    };
+
+    // Call original thread routine
+    let result = (thread_routine.start_routine)(thread_routine.args);
+
+    unsafe {
+        // FIXME: Potential memory leak. If the thread is cancelled we might not free allocated memory.
+        free(stack_addr)
+    }
+    result
+}
+
+/// This function intercepts the standard POSIX pthread_create `pthread_create` function and wraps it with custom
+/// YK functionality.
+///
+/// The function stores `start_routine` and `args` pointers in `ThreadRoutine` struct raw pointer.
+/// `ThreadRoutine` is passed to the `wrap_thread_routine` function as a raw pointer,
+/// that will use it to call the original routine with the original arguments.
+///
+/// The `ThreadRoutine` raw pointer is initialised by the `Box::into_raw` call.
+/// It will be dropped once it's re-constructed as a box from the raw pointer. i.e. when `Box::from_raw` is called.
```

Comment:
This sentence is correct but not useful, because it doesn't say *who* will call `Box::from_raw`.

Reply:
Added more info 👉 cfcd40160b15d79e6c0c2e7bf54951f620e92e3e